### PR TITLE
1126 multiprocess caching datasets PicklingError

### LINKED ISSUE #1126: Lightning Segmentation tutorial breaks on multiple GPU's (CacheDataset?)
Reproducing the spleen Lightning segmentation tutorial from [here](https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d_lightning.ipynb). Full logs of two errors (1) and (2) shared below. You can reproduce with full code from the notebook into a singly python file can be copied from (3)
With the latest library versions, and running on two GPU's results in the following error (1):
In multiprocessing/reduction.py > def dump:
image
After reading full log (1) I've commented out `CropForegroundd(keys=["image", "label"], source_key="image"),` from both `train_transforms` and `val_transforms` and I get the following error (2):
In multiprocessing/reduction.py > def dump:
image

(1) Full log:

> Exception has occurred: PicklingError
> Can't pickle <function CropForegroundd.<lambda> at 0x7fd6e5159b80>: attribute lookup CropForegroundd.<lambda> on monai.transforms.croppad.dictionary failed
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/reduction.py", line 60, in dump
>     ForkingPickler(file, protocol).dump(obj)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 47, in _launch
>     reduction.dump(process_obj, fp)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
>     self._launch(process_obj)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
>     super().__init__(process_obj)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
>     return Popen(process_obj)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/process.py", line 121, in start
>     self._popen = self._Popen(self)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 149, in start_processes
>     process.start()
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 200, in spawn
>     return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/site-packages/pytorch_lightning/accelerators/ddp_spawn_backend.py", line 64, in train
>     mp.spawn(self.ddp_train, nprocs=self.nprocs, args=(self.mp_queue, model,))
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 454, in fit
>     results = self.accelerator_backend.train()
>   File "/nfs/users/ext_bart.michiels/projects/examples/test_multi_gpu.py", line 213, in <module>
>     trainer.fit(net)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 97, in _run_module_code
>     _run_code(code, mod_globals, init_globals,
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 265, in run_path
>     return _run_module_code(code, init_globals, run_name,
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 194, in _run_module_as_main
>     return _run_code(code, main_globals, None,

(2) Full log:

> Exception has occurred: TypeError
> cannot pickle '_thread.lock' object
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/reduction.py", line 60, in dump
>     ForkingPickler(file, protocol).dump(obj)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 47, in _launch
>     reduction.dump(process_obj, fp)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
>     self._launch(process_obj)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
>     super().__init__(process_obj)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
>     return Popen(process_obj)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/multiprocessing/process.py", line 121, in start
>     self._popen = self._Popen(self)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 149, in start_processes
>     process.start()
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 200, in spawn
>     return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/site-packages/pytorch_lightning/accelerators/ddp_spawn_backend.py", line 64, in train
>     mp.spawn(self.ddp_train, nprocs=self.nprocs, args=(self.mp_queue, model,))
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 454, in fit
>     results = self.accelerator_backend.train()
>   File "/nfs/users/ext_bart.michiels/projects/examples/test_multi_gpu.py", line 213, in <module>
>     trainer.fit(net)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 97, in _run_module_code
>     _run_code(code, mod_globals, init_globals,
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 265, in run_path
>     return _run_module_code(code, init_globals, run_name,
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "/nfs/projects/dgx1-pod1-medical-project/anaconda/envs/covid_clf/lib/python3.8/runpy.py", line 194, in _run_module_as_main
>     return _run_code(code, main_globals, None,

(3) Full python code from the notebook to reproduce the errors:

> import glob
> import os
> import shutil
> import tempfile
> import matplotlib.pyplot as plt
> import pytorch_lightning
> import torch
> from monai.apps import download_and_extract
> from monai.config import print_config
> from monai.data import CacheDataset, list_data_collate
> from monai.inferers import sliding_window_inference
> from monai.losses import DiceLoss
> from monai.metrics import compute_meandice
> from monai.networks.layers import Norm
> from monai.networks.nets import UNet
> from monai.transforms import (
>     AddChanneld,
>     Compose,
>     CropForegroundd,
>     LoadNiftid,
>     Orientationd,
>     RandCropByPosNegLabeld,
>     ScaleIntensityRanged,
>     Spacingd,
>     ToTensord,
> )
> from monai.utils import set_determinism
> print_config()
> directory = os.environ.get("MONAI_DATA_DIRECTORY")
> root_dir = tempfile.mkdtemp() if directory is None else directory
> print(root_dir)
> resource = "https://drive.google.com/uc?id=1jzeNU1EKnK81PyTsrx0ujfNl-t0Jo8uE"
> md5 = "410d4a301da4e5b2f6f86ec3ddba524e"
> compressed_file = os.path.join(root_dir, "Task09_Spleen.tar")
> data_dir = os.path.join(root_dir, "Task09_Spleen")
> if not os.path.exists(data_dir):
>     download_and_extract(resource, compressed_file, root_dir, md5)
> class Net(pytorch_lightning.LightningModule):
>     def __init__(self):
>         super().__init__()
>         self._model = UNet(
>             dimensions=3,
>             in_channels=1,
>             out_channels=2,
>             channels=(16, 32, 64, 128, 256),
>             strides=(2, 2, 2, 2),
>             num_res_units=2,
>             norm=Norm.BATCH,
>         )
>         self.loss_function = DiceLoss(to_onehot_y=True, softmax=True)
>         self.best_val_dice = 0
>         self.best_val_epoch = 0
> 
>     def forward(self, x):
>         return self._model(x)
> 
>     def prepare_data(self):
>         # set up the correct data path
>         train_images = sorted(glob.glob(os.path.join(data_dir, "imagesTr", "*.nii.gz")))
>         train_labels = sorted(glob.glob(os.path.join(data_dir, "labelsTr", "*.nii.gz")))
>         data_dicts = [
>             {"image": image_name, "label": label_name}
>             for image_name, label_name in zip(train_images, train_labels)
>         ]
>         train_files, val_files = data_dicts[:-9], data_dicts[-9:]
> 
>         # set deterministic training for reproducibility
>         set_determinism(seed=0)
> 
>         # define the data transforms
>         train_transforms = Compose(
>             [
>                 LoadNiftid(keys=["image", "label"]),
>                 AddChanneld(keys=["image", "label"]),
>                 Spacingd(
>                     keys=["image", "label"], pixdim=(1.5, 1.5, 2.0), mode=("bilinear", "nearest"),
>                 ),
>                 Orientationd(keys=["image", "label"], axcodes="RAS"),
>                 ScaleIntensityRanged(
>                     keys=["image"], a_min=-57, a_max=164, b_min=0.0, b_max=1.0, clip=True,
>                 ),
>                 #CropForegroundd(keys=["image", "label"], source_key="image"),
>                 # randomly crop out patch samples from big image based on pos / neg ratio
>                 # the image centers of negative samples must be in valid image area
>                 RandCropByPosNegLabeld(
>                     keys=["image", "label"],
>                     label_key="label",
>                     spatial_size=(96, 96, 96),
>                     pos=1,
>                     neg=1,
>                     num_samples=4,
>                     image_key="image",
>                     image_threshold=0,
>                 ),
>                 # user can also add other random transforms
>                 # RandAffined(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=1.0, spatial_size=(96, 96, 96),
>                 #             rotate_range=(0, 0, np.pi/15), scale_range=(0.1, 0.1, 0.1)),
>                 ToTensord(keys=["image", "label"]),
>             ]
>         )
>         val_transforms = Compose(
>             [
>                 LoadNiftid(keys=["image", "label"]),
>                 AddChanneld(keys=["image", "label"]),
>                 Spacingd(
>                     keys=["image", "label"], pixdim=(1.5, 1.5, 2.0), mode=("bilinear", "nearest"),
>                 ),
>                 Orientationd(keys=["image", "label"], axcodes="RAS"),
>                 ScaleIntensityRanged(
>                     keys=["image"], a_min=-57, a_max=164, b_min=0.0, b_max=1.0, clip=True,
>                 ),
>                 #CropForegroundd(keys=["image", "label"], source_key="image"),
>                 ToTensord(keys=["image", "label"]),
>             ]
>         )
> 
>         # we use cached datasets - these are 10x faster than regular datasets
>         self.train_ds = CacheDataset(
>             data=train_files, transform=train_transforms, cache_rate=1.0, num_workers=4,
>         )
>         self.val_ds = CacheDataset(
>             data=val_files, transform=val_transforms, cache_rate=1.0, num_workers=4,
>         )
>         # self.train_ds = monai.data.Dataset(data=train_files, transform=train_transforms)
>         # self.val_ds = monai.data.Dataset(data=val_files, transform=val_transforms)
> 
>     def train_dataloader(self):
>         train_loader = torch.utils.data.DataLoader(
>             self.train_ds, batch_size=2, shuffle=True, num_workers=4, collate_fn=list_data_collate,
>         )
>         return train_loader
> 
>     def val_dataloader(self):
>         val_loader = torch.utils.data.DataLoader(self.val_ds, batch_size=1, num_workers=4)
>         return val_loader
> 
>     def configure_optimizers(self):
>         optimizer = torch.optim.Adam(self._model.parameters(), 1e-4)
>         return optimizer
> 
>     def training_step(self, batch, batch_idx):
>         images, labels = batch["image"], batch["label"]
>         output = self.forward(images)
>         loss = self.loss_function(output, labels)
>         tensorboard_logs = {"train_loss": loss.item()}
>         return {"loss": loss, "log": tensorboard_logs}
> 
>     def validation_step(self, batch, batch_idx):
>         images, labels = batch["image"], batch["label"]
>         roi_size = (160, 160, 160)
>         sw_batch_size = 4
>         outputs = sliding_window_inference(images, roi_size, sw_batch_size, self.forward)
>         loss = self.loss_function(outputs, labels)
>         value = compute_meandice(
>             y_pred=outputs,
>             y=labels,
>             include_background=False,
>             to_onehot_y=True,
>             mutually_exclusive=True,
>         )
>         return {"val_loss": loss, "val_dice": value}
> 
>     def validation_epoch_end(self, outputs):
>         val_dice, val_loss, num_items = 0, 0, 0
>         for output in outputs:
>             val_dice += output["val_dice"].sum().item()
>             val_loss += output["val_loss"].sum().item()
>             num_items += len(output["val_dice"])
>         mean_val_dice = torch.tensor(val_dice / num_items)
>         mean_val_loss = torch.tensor(val_loss / num_items)
>         tensorboard_logs = {
>             "val_dice": mean_val_dice,
>             "val_loss": mean_val_loss,
>         }
>         if mean_val_dice > self.best_val_dice:
>             self.best_val_dice = mean_val_dice
>             self.best_val_epoch = self.current_epoch
>         print(
>             f"current epoch: {self.current_epoch} current mean dice: {mean_val_dice:.4f}"
>             f"\nbest mean dice: {self.best_val_dice:.4f} at epoch: {self.best_val_epoch}"
>         )
>         return {"log": tensorboard_logs}
> net = Net()
> log_dir = os.path.join(root_dir, "logs")
> tb_logger = pytorch_lightning.loggers.TensorBoardLogger(
>     save_dir=log_dir
> )
> checkpoint_callback = pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint(
>     filepath=os.path.join(log_dir, "{epoch}-{val_loss:.2f}-{val_dice:.2f}")
> )
> 
> trainer = pytorch_lightning.Trainer(
>     gpus=[0,1],
>     max_epochs=600,
>     logger=tb_logger,
>     checkpoint_callback=checkpoint_callback,
>     num_sanity_val_steps=1,
> )
> 
> trainer.fit(net)
> 
> print(f"train completed, best_metric: {net.best_val_dice:.4f} at epoch {net.best_val_epoch}")

### PR DESCRIPTION
**Pull request by wyli**
Fixes #1126

### Description
- load cache lazily to avoid pickling complex objects (lmdbdataset)
- refactored tqdm calls for cachedataset (this fixes 1126)
- this PR mainly changes the caching behaviour but doesn't change the API

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.

### CODE REVIEW

**Review by Nic-Ma — APPROVED** [review]
Looks good to me.
Thanks.
